src/protego.py

```diff
@@ -0,0 +1,337 @@
+        if path.endswith('/index.html'):
+            self.allow(path[:-10] + '$')
+
+        # Sort rules based on length. If two disallow & allow rules have same
+        # length, allow rule is prioritised.
+        self._rules.sort(key=lambda r: (len(r.value), r.field == 'allow'), reverse=True)
+
+    def disallow(self, path):
+        path = self._quote_path(path, safe='/*$')
+        self._rules.append(Rule(field='disallow', value=path))
+
+        # Sort rules based on length. If two disallow & allow rules have same
+        # length, allow rule is prioritised.
+        self._rules.sort(key=lambda r: (len(r.value), r.field == 'allow'), reverse=True)
+
+    def _parser_match(self, pattern, url):
+        """Return True if pattern matches the URL, otherwise return False."""
+        pattern = re.sub(r'\*+', '*', pattern)
+        s = re.split(r'([$*])', pattern)
+        for index, substr in enumerate(s):
+            if substr not in ['*', '$']:
+                s[index] = re.escape(substr)
+            else:
+                s[index] = s[index].replace('*', '.*')
+        pattern = '^' + ''.join(s)
+        if re.findall(pattern, url):
+            return True
+        return False
+
+    def can_fetch(self, url):
+        """Return if the url can be fetched."""
+        url = self._quote_path(url)
+        allowed = True
+        for rule in self._rules:
+            if self._parser_match(rule.value, url):
+                if rule.field == 'disallow':
+                    allowed = False
+                break
+        return allowed
+
+    @property
+    def crawl_delay(self):
+        """Get & set crawl delay for the rule set."""
+        return self._crawl_delay
+
+    @crawl_delay.setter
+    def crawl_delay(self, delay):
+        try:
+            delay = float(delay)
+        except ValueError:
+            delay = None
+        self._crawl_delay = delay
+
+    @property
+    def request_rate(self):
+        """Get & set request rate for the rule set."""
+        return self._req_rate
+
+    @request_rate.setter
+    def request_rate(self, value):
+        try:
+            parts = value.split()
+            if len(parts) == 2:
+                rate, time_period = parts
+            else:
+                rate, time_period = parts[0], ''
+
+            requests, seconds = rate.split('/')
+            requests = int(requests)
+            time_unit = seconds[-1].lower()
+            seconds = int(seconds[:-1])
+            if time_unit == 'm':
+                seconds *= 60
+            elif time_unit == 'h':
+                seconds *= 3600
+            elif time_unit == 'd':
+                seconds *= 86400
+
+            start_time = None
+            end_time = None
+            if time_period:
+                start_time, end_time = time_period.split('-')
+                start_time = time(int(start_time[:2]), int(start_time[-2:]))
+                end_time = time(int(end_time[:2]), int(end_time[-2:]))
+        except:
```

Comment:
```suggestion
        except Exception:
```

This is so that you do not catch exceptions that are not meant to be caught, like https://docs.python.org/3/library/exceptions.html#KeyboardInterrupt